### ISSUE TITLE
[Flang] Incorrect diagnose when proc-iface in a procedure declaration statement is a function that returns a polymorphic procedure pointer

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#83292]
Consider the following test case
```
  MODULE M

    TYPE :: DT
      real :: r
    END TYPE

    INTERFACE
      FUNCTION FunI()
        IMPORT
        CLASS(DT), POINTER :: FunI
      END FUNCTION
    END INTERFACE

    INTERFACE
      FUNCTION FunRetPtr1()
        IMPORT
        PROCEDURE(FunI), POINTER :: FunRetPtr1
      END FUNCTION
    END INTERFACE
  END MODULE


  PROGRAM PrtAssignTarExpr
  USE M
  IMPLICIT NONE

  PROCEDURE(FunRetPtr1), POINTER :: Ptr1
  END
```

Flang currently issues an error:
```
./t.f:27:37: error: CLASS entity 'ptr1' must be a dummy argument, allocatable, or object pointer
    PROCEDURE(FunRetPtr1), POINTER :: Ptr1
                                      ^^^^
 ```
 
 The code seems conforming to me and the error message doesn't quite make sense to me.

**Comment by klausler** [comment]
https://github.com/llvm/llvm-project/pull/83733

**Comment by DanielCChen** [comment]
I have verified the reproducer and the original test case are fixed by PR #83733.
